src/serial.rs

```diff
@@ -0,0 +1,484 @@
+///
+/// // We expect, in order,
+/// // 1. A read that producers 0x23,
+/// // 2. A write of [0x55, 0xAA]
+/// // 3. A flush
+/// let transactions = [
+///     Transaction::read(0x23),
+///     Transaction::write([0x55, 0xAA]),
+///     Transaction::flush()
+/// ];
+///
+/// let serial = Mock::new(&transactions);
+/// ```
+pub struct Transaction<Word> {
+    /// A collection of modes
+    ///
+    /// Since we need to express a blocking write in terms of
+    /// multiple writes, we aggregate all of them into this
+    /// member. Then, they are handed-off to the mock on
+    /// construction.
+    mode: Vec<Mode<Word>>,
+}
+
+impl<Word> Transaction<Word>
+where
+    Word: Clone,
+{
+    /// Expect a serial read that returns the expected word
+    pub fn read(expected: Word) -> Self {
+        Transaction {
+            mode: vec![Mode::Read(expected)],
+        }
+    }
+
+    /// Expect a serial write that transmits the specified word
+    ///
+    /// `write()` accepts a collection of words. If you're expecting
+    /// a blocking write, you may want to supply a slice of words that
+    /// would be used in the write, rather than listing them out one-by-
+    /// one. Or, if you're expecting multiple, non-blocking writes, you
+    /// may choose to aggregate them into a slice instead of listing them
+    /// out one-by-one.
+    pub fn write<Ws>(words: Ws) -> Self
+    where
+        Ws: AsRef<[Word]>,
+    {
+        Transaction {
+            mode: words.as_ref().iter().cloned().map(Mode::Write).collect(),
+        }
+    }
+
+    /// Expecte a caller to flush the serial buffers
+    pub fn flush() -> Self {
+        Transaction {
+            mode: vec![Mode::Flush],
+        }
+    }
+}
+
+/// Mock serial device
+///
+/// The mock serial device can be loaded with expected transactions, then
+/// passed-on into a serial device user. If the expectations were not met
+/// in the specified order, the type causes a panic and describes what
+/// expectation wasn't met.
+///
+/// The type is clonable so that it may be shared with a serial
+/// device user. Under the hood, both cloned mocks will share
+/// the same state, allowing your handle to eventually call `done()`,
+/// if desired.
+pub struct Mock<'a, Word> {
+    /// The expected operations upon the mock
+    ///
+    /// It's in an arc to maintain shared state, and in a mutex
+    /// to make it thread safe. It's then wrapped in an `Option`
+    /// so that we can take it in the call to `done()`.
+    expected_modes: Arc<Mutex<Option<VecDeque<Mode<Word>>>>>,
+    /// We'll note that this phantom type and lifetime is actually
+    /// unnecessary. As noted above, we try to keep a consistent
+    /// public API with the existing spi and i2c modules. Given that
+    /// this lifetime is necessary for them, it is also necessary here.
+    ///
+    /// Of course, this could quickly be removed if the overall public
+    /// API changes.
+    marker: PhantomData<&'a Word>,
```

Comment:
Hm, what do you mean with "Given that this lifetime is necessary for them, it is also necessary here"?

If the marker / lifetime is not required, then you can remove it. However the lifetime is used below, right?

Reply:
Sorry I wasn't clear. The goal is to maintain API consistency. Both a SPI and I2C mock each require a lifetime in the type signature, since they're based on the generic mock implementation. We add a lifetime here only to keep the APIs consistent.

The phantom marker is meaningless; we would compile fine without it. Should we consider breaking this consistency?

Reply:
Yeah, I wouldn't include a lifetime parameter that isn't actually used / needed. In practice you don't explicitly pass in a lifetime anyways, so for the user the absence of the lifetime parameter doesn't break API consistency.

Reply:
You got it. Removed in f426d65.